File: api-quarkus/src/test/java/br/ufrn/musi/BuscarObrasTest.java
package br.ufrn.musi;

import br.ufrn.musi.aplicacao.BuscarObras;
import br.ufrn.musi.aplicacao.FonteDeObras;
import br.ufrn.musi.dominio.Dominio.*;
import org.junit.jupiter.api.Test;
import java.util.List;

import static br.ufrn.musi.dominio.Dominio.satisfaz;
import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Repare que não há `@QuarkusTest`: é um teste de unidade comum, porque a classe
 * sob teste não depende de framework.
 */
class BuscarObrasTest {

    static final List<Obra> ACERVO = DominioTest.ACERVO;

    final FonteDeObras emMemoria = filtro ->
        ACERVO.stream().filter(o -> satisfaz(o, filtro)).toList();

    @Test void buscaPorRitmo() {
        var obras = new BuscarObras(emMemoria).executar(new Tem("ritmo", "baiao"));
        assertEquals(List.of("obra-03", "obra-05"), obras.stream().map(Obra::id).toList());
    }

    @Test void buscaComposta() {
        var filtro = new E(List.of(new Tem("genero", "mpb"), new Ate(1970)));
        assertEquals(2, new BuscarObras(emMemoria).executar(filtro).size());
    }
}

File: api-ktor/src/test/kotlin/br/ufrn/musi/BuscarObrasTest.kt
package br.ufrn.musi

import br.ufrn.musi.aplicacao.BuscarObras
import br.ufrn.musi.aplicacao.FonteDeObras
import br.ufrn.musi.dominio.*
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * O ganho concreto da porta: este teste não sobe servidor, não abre porta e não usa
 * mock de biblioteca HTTP. Substitui a INTERFACE por seis linhas.
 */
class BuscarObrasTest {

    private val acervo = listOf(
        Obra("obra-01", "Ponteio", "Edu Lobo", 1967,
            listOf(Faceta("genero", "mpb"), Faceta("ritmo", "ponteio"))),
        Obra("obra-03", "Asa Branca", "Luiz Gonzaga", 1947,
            listOf(Faceta("genero", "forro"), Faceta("ritmo", "baiao"))),
    )

    private val emMemoria = object : FonteDeObras {
        override suspend fun buscar(filtro: Filtro) = acervo.filter { it.satisfaz(filtro) }
    }

    @Test
    fun buscaPorRitmo() = runTest {
        val obras = BuscarObras(emMemoria)(Filtro.Tem("ritmo", "baiao"))
        assertEquals(listOf("obra-03"), obras.map { it.id })
    }

    @Test
    fun buscaComposta() = runTest {
        val filtro = Filtro.E(listOf(Filtro.Tem("genero", "mpb"), Filtro.Ate(1970)))
        assertEquals(1, BuscarObras(emMemoria)(filtro).size)
    }
}
